File: dfmux/python/Housekeeping.py
from spt3g import core
from spt3g.dfmux import DfMuxHousekeepingMap, HkBoardInfo, HkMezzanineInfo, HkModuleInfo, HkChannelInfo, DfMuxWiringMap, DfMuxChannelMapping

from .TuberClient import TuberClient
import socket, struct, time
import numpy

@core.indexmod
class HousekeepingConsumer(object):
    '''
    Collect housekeeping data from the mux boards defined in the wiring map.
    Will add a key called 'DfMuxHousekeeping' to any housekeeping frame that
    goes by containing the data as of the arrival of the housekeeping frame.
    Use in conjunction with a dfmux.PeriodicHousekeepingCollector to get
    data at fixed intervals.

    Also emits a Wiring frame from the housekeeping data.  This requires a
    recent (as of November 2018) version of pydfmux in order to read mapped
    channel names from each board.  If ignore_wiring=False, assumes that
    the wiring map is handled by a separate process.

    If collecting real-time data, you may want to set subprocess=True when
    adding this module.
    '''
    def __init__(self, ignore_wiring=False):
        self.tuber = {}
        self.board_map = {}
        self.board_serials = []
        self.buffer = []
        self.hwmf = None
        self.ignore_wiring = ignore_wiring

    def map_boards(self):
        '''
        Cache IP address, crate serial and slot number and
        create a TuberClient for each IceBoard.
        '''
        boards = [b for b in self.board_serials if b not in self.board_map]
        if not boards:
            return

        ips = {}
        crates = {}
        slots = {}

        # create tubers for each missing board
        for board in boards:
            ip = socket.inet_aton(socket.gethostbyname('iceboard{}.local'.format(board)))
            ips[board] = struct.unpack("i", ip)[0]
            self.tuber[board] = TuberClient(socket.inet_ntoa(ip), timeout=5.0)

        # get crate ID for each missing board
        for board in boards:
            self.tuber[board].CallMethod('Dfmux', '_get_backplane_serial', False)
        for board in boards:
            reply = self.tuber[board].GetReply()[0]
            if reply.get('error', None):
                crates[board] = -1
            else:
                crates[board] = int(reply['result'])

        # get slot ID for each missing board
        for board in boards:
            if crates[board] >= 0:
                self.tuber[board].CallMethod('Dfmux', 'get_backplane_slot', False)
        for board in boards:
            if crates[board] < 0:
                slots[board] = -1
            else:
                reply = self.tuber[board].GetReply()[0]
                if reply.get('error', None):
                    slots[board] = -1
                else:
                    slots[board] = int(reply['result'])

        # store
        for board in boards:
            self.board_map[board] = (ips[board], crates[board], slots[board])

    def __call__(self, frame):

        # If boards have been mapped already, then process every frame
        # as received.
        if len(self.board_map):
            return self.ProcessBuffered(frame)

        # Otherwise, process at least one Timepoint frame first to gather
        # a list of IceBoards for which housekeeping data are required,
        # Then process any buffered frames after Timepoint frame,
        # but emit before the Timepoint
        if frame.type == core.G3FrameType.Timepoint:
            tp_frames = self.ProcessBuffered(frame)
            frames = sum([self.ProcessBuffered(fr) for fr in self.buffer], [])
            frames += tp_frames
            self.buffer = []
            return frames

        # Buffer until a Timepoint frame is received
        self.buffer.append(frame)
        return False

    def ProcessBuffered(self, frame):
        '''
        Process frames in the buffered order.  Returned value should
        always be a list of frames, possibly empty.
        '''
        if frame.type == core.G3FrameType.Timepoint:
            self.board_serials = [('%04d' % k) for k in frame['DfMux'].keys()]
            return [frame]

        if frame.type == core.G3FrameType.Wiring:
            if self.ignore_wiring:
                core.log_warn(
                    "Received a wiring frame, which may be inconsistent with "
                    "board housekeeping data.  Store mapped channel names on the "
                    "board using updated pydfmux/hidfmux software.",
                    unit="HousekeepingConsumer",
                )
            else:
                core.log_fatal(
                    "Received spurious wiring frame.  Do not use "
                    "PyDfMuxHardwareMapInjector with the HousekeepingConsumer.  "
                    "You may update pydfmux/hidfmux to a newer version that "
                    "stores mapped channel names on the boards, and rearrange "
                    "your data acquisition script.",
                    unit="HousekeepingConsumer",
                )

        if frame.type == core.G3FrameType.Housekeeping:
            self.map_boards()

            hwm = DfMuxWiringMap()
            hkdata = DfMuxHousekeepingMap()
            try:
                for board in self.board_serials:
                    self.tuber[board].CallMethod('Dfmux', '_dump_housekeeping', False)
                    time.sleep(0.02) # Stagger return data transfer a little to
                                     # avoid overloading the network on the return

                found = False
                ismkid = False
                for board in self.board_serials:
                    dat = self.tuber[board].GetReply()[0]['result']

                    boardhk = self.HousekeepingFromJSON(dat)
                    if boardhk.firmware_name:
                        ismkid = ismkid or "mkid" in boardhk.firmware_name.lower()
                    hkdata[int(boardhk.serial)] = boardhk

                    if self.ignore_wiring:
                        continue

                    ip, crate, slot = self.board_map[board]
                    boardw = self.WiringFromJSON(dat, ip, crate, slot)
                    for key in boardw.keys():
                        hwm[key] = boardw[key]
                        found = True

                if not found and not self.ignore_wiring:
                    core.log_fatal("No mapped channels found on any IceBoards. "
                                   "You may need to update pydfmux to a newer version of pydfmux "
                                   "that stores mapped channel names on the boards, and reload "
                                   "the hardware map.",
                                   unit='HousekeepingConsumer')

                frame['DfMuxHousekeeping'] = hkdata

                if self.ignore_wiring:
                    return [frame]

                hwmf = core.G3Frame(core.G3FrameType.Wiring)
                hwmf['WiringMap'] = hwm
                hwmf['ReadoutSystem'] = 'RF-ICE' if ismkid else 'ICE'

                if self.hwmf is None:
                    self.hwmf = hwmf
                    # If this is the first time the consumer is triggered, make sure
                    # a Housekeeping and WiringMap frame are issued before any
                    # Timepoint frames
                    frames = [hwmf, frame]
                    return frames
                else:
                    # Compare wiring maps and re-issue frame if anything has changed
                    old_hwm = self.hwmf['WiringMap']
                    if (set(hwm.keys()) ^ set(old_hwm.keys())):
                        self.hwmf = hwmf
                        return [hwmf, frame]
                    for k in hwm.keys():
                        try:
                            if vars(hwm[str(k)]) != vars(old_hwm[str(k)]):
                                self.hwmf = hwmf
                                return [hwmf, frame]
                        except:
                            core.log_error("Invalid HWM key %r" % k)

                # If we get here then the wiring map hasn't changed,
                # so return the populated Housekeeping frame as it is
                return [frame]

            except socket.timeout:
                core.log_error('Timeout collecting housekeeping data from mux boards. Dropping housekeeping sample', unit='HousekeepingConsumer')
                return []
            except Exception as e:
                core.log_error('Error (%s) collecting housekeeping data from mux boards. Dropping housekeeping sample' % e, unit='HousekeepingConsumer')
                return []

        return [frame]

    @classmethod
    def HousekeepingFromJSON(cls, dat):
        '''
        Build HKBoardInfo object from a JSON blob returned by the
        _dump_housekeeping call
        '''
        # Board-global quantities
        boardhk = HkBoardInfo()
        if 'is128x' in dat:
            boardhk.is128x = dat['is128x']
        else:
            boardhk.is128x = False
        year = dat['timestamp']['y']
        if year == 0:
            # It probably isn't 1900
            systime = time.gmtime()
            # Check for New Year's, assuming no more than 24 hours clock slew
            if dat['timestamp']['d'] == 1 and systime.tm_yday >= 365:
                year = systime.tm_year + 1
            elif dat['timestamp']['d'] >= 365 and systime.tm_yday == 1:
                year = systime.tm_year - 1
            else:
                year = systime.tm_year
            year -= 2000
        boardhk.timestamp = core.G3Time(y=year,d=dat['timestamp']['d'],h=dat['timestamp']['h'],m=dat['timestamp']['m'],s=dat['timestamp']['s'],ss=dat['timestamp']['ss'])

        boardhk.timestamp_port = str(dat['timestamp_port'])
        boardhk.serial = str(dat['serial'])
        if 'firmware_name' in dat:
            boardhk.firmware_name = str(dat['firmware_name'])
            boardhk.firmware_version = str(dat['firmware_version'])
        boardhk.fir_stage = dat['fir_stage']
        for i in dat['currents'].items():
            boardhk.currents[str(i[0])] = i[1]
        for i in dat['voltages'].items():
            boardhk.voltages[str(i[0])] = i[1]
        for i in dat['temperatures'].items():
            boardhk.temperatures[str(i[0])] = i[1]

        # Mezzanines
        for n, mezz in enumerate(dat['mezzanines']):
            mezzhk = HkMezzanineInfo()
            mezzhk.present = mezz['present']
            mezzhk.power = mezz['power']
            if mezzhk.present:
                if 'ipmi' in mezz:
                    mezzhk.serial = str(mezz['ipmi']['product']['serial_number'])
                    mezzhk.part_number = str(mezz['ipmi']['product']['part_number'])
                    mezzhk.revision = str(mezz['ipmi']['product']['version_number'])
                if 'currents' in mezz:
                    for i in mezz['currents'].items():
                        mezzhk.currents[str(i[0])] = i[1]
                if 'voltages' in mezz:
                    for i in mezz['voltages'].items():
                        mezzhk.voltages[str(i[0])] = i[1]

            if mezzhk.present and mezzhk.power:
                if 'temperature' in mezz:
                    mezzhk.temperature = mezz['temperature']
                # these parameters are not in the 64x housekeeping tuber
                mezzhk.squid_heater = mezz.get('squid_heater', 0.0)
                mezzhk.squid_controller_power = mezz.get('squid_controller_power', False)
                mezzhk.squid_controller_temperature = mezz.get('squid_controller_temperature', 0.0)

            # Modules
            for m, mod in enumerate(mezz['modules']):
                modhk = HkModuleInfo()
                modhk.routing_type = str(mod['routing'][0])
                modhk.module_number = m+1
                if mezzhk.present and mezzhk.power:
                    if 'gains' in mod:
                        modhk.carrier_gain = mod['gains']['carrier']
                        modhk.nuller_gain = mod['gains']['nuller']
                    if 'overload' in mod:
                        modhk.carrier_railed = mod['overload']['carrier']
                        modhk.nuller_railed = mod['overload']['nuller']
                        modhk.demod_railed = mod['overload']['demod']
                    if 'squid_current_bias' in mod:
                        modhk.squid_current_bias = mod['squid_current_bias']
                    if 'squid_flux_bias' in mod:
                        modhk.squid_current_bias = mod['squid_flux_bias']
                    if 'squid_feedback' in mod:
                        modhk.squid_feedback = str(mod['squid_feedback'])
                    if 'nco_frequency' in mod:
                        modhk.nco_frequency = mod['nco_frequency'] * core.G3Units.Hz

                if 'squid_tuning' in mod and mod['squid_tuning'] is not None:
                    modhk.squid_state = str(mod['squid_tuning']['state'])
                    modhk.squid_transimpedance = mod['squid_tuning']['transimpedance'] if mod['squid_tuning']['transimpedance'] is not None else numpy.nan
                    modhk.squid_p2p = mod['squid_tuning']['p2p'] if mod['squid_tuning']['p2p'] is not None else numpy.nan

                for k, chan in enumerate(mod['channels']):
                    chanhk = HkChannelInfo()
                    chanhk.channel_number = k+1
                    chanhk.carrier_amplitude = chan['carrier_amplitude']
                    chanhk.nuller_amplitude = chan['nuller_amplitude']
                    if 'dan_gain' in chan:
                        chanhk.dan_gain = chan['dan_gain']
                    if 'dan_streaming_enable' in chan:
                        chanhk.dan_streaming_enable = chan['dan_streaming_enable']
                    if 'frequency' in chan:
                        chanhk.carrier_frequency = chan['frequency'] * core.G3Units.Hz
                        chanhk.demod_frequency = chan['frequency'] * core.G3Units.Hz
                    else:
                        chanhk.carrier_frequency = chan['carrier_frequency'] * core.G3Units.Hz
                        chanhk.demod_frequency = chan['demod_frequency'] * core.G3Units.Hz
                    if 'carrier_phase' in chan:
                        chanhk.carrier_phase = chan['carrier_phase'] * core.G3Units.deg
                        chanhk.nuller_phase = chan['nuller_phase'] * core.G3Units.deg
                        chanhk.demod_phase = chan['demod_phase'] * core.G3Units.deg
                    if 'dan_accumulator_enable' in chan:
                        chanhk.dan_accumulator_enable = chan['dan_accumulator_enable']
                    if 'dan_feedback_enable' in chan:
                        chanhk.dan_feedback_enable = chan['dan_feedback_enable']
                    if 'dan_railed' in chan:
                        chanhk.dan_railed = chan['dan_railed']
                    if 'tuning' in chan and chan['tuning'] is not None:
                        chanhk.state = str(chan['tuning']['state'])
                        if ('rlatched' in chan['tuning'] and 
                            chan['tuning']['rlatched'] is not None):
                            chanhk.rlatched = chan['tuning']['rlatched']
                        if ('rnormal' in chan['tuning'] and 
                            chan['tuning']['rnormal'] is not None):
                            chanhk.rnormal = chan['tuning']['rnormal']
                        if ('rfrac_achieved' in chan['tuning'] and 
                            chan['tuning']['rfrac_achieved'] is not None):
                            chanhk.rfrac_achieved = chan['tuning']['rfrac_achieved']
                        if ('loopgain' in chan['tuning'] and
                            chan['tuning']['loopgain'] is not None):
                            chanhk.loopgain = chan['tuning']['loopgain']

                    modhk.channels[k+1] = chanhk
                mezzhk.modules[m+1] = modhk
            boardhk.mezz[n+1] = mezzhk
    
        return boardhk

    @classmethod
    def WiringFromJSON(cls, dat, ip, crate, slot):
        '''
        Build WiringMap object from a JSON blob returned by the
        _dump_housekeeping call
        '''
        found = False
        hwm = DfMuxWiringMap()
        serial = int(dat['serial'])
        for imezz, mezz in enumerate(dat['mezzanines']):
            for imod, mod in enumerate(mezz['modules']):
                module = imod + len(mezz['modules']) * imezz
                for ichan, chan in enumerate(mod['channels']):
                    name = (chan.get('tuning', {}) or {}).get('name', None)
                    if not name:
                        continue
                    mapping = DfMuxChannelMapping()
                    mapping.board_ip = ip
                    mapping.board_serial = serial
                    mapping.board_slot = slot
                    mapping.crate_serial = crate
                    mapping.module = module
                    mapping.channel = ichan
                    try:
                        name = str(name)
                        hwm[name] = mapping
                        found = True
                    except:
                        core.log_error("Invalid channel name %r" % (name))
        if not found:
            core.log_error("No mapped channels found on iceboard%04d. "
                           "You may need to update pydfmux to a newer version of pydfmux "
                           "that stores mapped channel names on the boards, and reload "
                           "the hardware map." % (serial),
                           unit='HousekeepingConsumer')
        return hwm

    @classmethod
    def HousekeepingForBoard(cls, hostname):
        '''
        Return HkBoardInfo object from a board. Useful for debugging.
        '''
        t = TuberClient(hostname)
        data = t.CallMethod('Dfmux', '_dump_housekeeping')
        return cls.HousekeepingFromJSON(data[0]['result'])

@core.indexmod
class PeriodicHousekeepingCollector(object):
    '''Inserts housekeeping frames every N timepoints.'''
    def __init__(self, N=15200):
        self.N = N
        self.count = 0
    def __call__(self, frame):
        ret = []
        if frame.type == core.G3FrameType.Timepoint:
            if self.count % self.N == 0:
                ret.append(core.G3Frame(core.G3FrameType.Housekeeping))
            self.count += 1
            ret.append(frame)
            return ret

@core.usefulfunc
def HousekeepingForBolo(hkmap, wiringmap, bolo, all_hk=False):
    '''
    Obtain the channel housekeeping information for a bolometer named "bolo"
    using the passed housekeeping and wiring maps.

    If all_hk is True, returns a tuple of the (board, mezz, module, channel)
    HK data instead of just the channel.
    '''

    wiringentry = wiringmap[bolo]
    board = hkmap[wiringentry.board_serial]
    mezz = board.mezz[(wiringentry.module // len(board.mezz[1].modules)) + 1]
    mod = mezz.modules[(wiringentry.module % len(mezz.modules)) + 1]
    chan = mod.channels[wiringentry.channel + 1]
    if all_hk:
        return (board, mezz, mod, chan)
    else:
        return chan

File: dfmux/src/Housekeeping.cxx
#include <sstream>
#include <pybindings.h>
#include <dfmux/Housekeeping.h>

#include <container_pybindings.h>

std::string HkChannelInfo::Description() const
{
	std::ostringstream s;
	s << "Channel " << channel_number << ", " <<
	    carrier_frequency/G3Units::MHz << " MHz (tuning: " << state << ")";
	return s.str();
}

template <class A> void HkChannelInfo::serialize(A &ar, unsigned v)
{
	using namespace cereal;

	G3_CHECK_VERSION(v);

	ar & make_nvp("G3FrameObject", base_class<G3FrameObject>(this));
	ar & make_nvp("channel_number", channel_number);
	ar & make_nvp("carrier_amplitude", carrier_amplitude);
	ar & make_nvp("carrier_frequency", carrier_frequency);
	ar & make_nvp("dan_accumulator_enable", dan_accumulator_enable);
	ar & make_nvp("dan_feedback_enable", dan_feedback_enable);
	ar & make_nvp("dan_streaming_enable", dan_streaming_enable);
	ar & make_nvp("dan_gain", dan_gain);
	ar & make_nvp("demod_frequency", demod_frequency);
	ar & make_nvp("nuller_amplitude", nuller_amplitude);
	ar & make_nvp("dan_railed", dan_railed);

	if (v >= 2) {
		ar & make_nvp("state", state);
		ar & make_nvp("rlatched", rlatched);
		ar & make_nvp("rnormal", rnormal);
		ar & make_nvp("rfrac_achieved", rfrac_achieved);
	}

	if (v == 3) {
		// Absorb a short-lived calibration constant that lived in
		// the HK data to address software limitations in Lyrebird.
		double res_conversion_factor;
		ar & make_nvp("res_conversion_factor", res_conversion_factor);
	}

	if (v > 4) {
		ar & make_nvp("loopgain", loopgain);
	}
}

std::string HkModuleInfo::Description() const
{
	std::ostringstream s;
	s << "Module " << module_number << " (SQUID: " << squid_state << ")";
	return s.str();
}

template <class A> void HkModuleInfo::serialize(A &ar, unsigned v)
{
	using namespace cereal;

	G3_CHECK_VERSION(v);

	ar & make_nvp("G3FrameObject", base_class<G3FrameObject>(this));
	ar & make_nvp("module_number", module_number);
	ar & make_nvp("carrier_gain", carrier_gain);
	ar & make_nvp("nuller_gain", nuller_gain);
	ar & make_nvp("demod_gain", demod_gain);
	ar & make_nvp("carrier_railed", carrier_railed);
	ar & make_nvp("nuller_railed", nuller_railed);
	ar & make_nvp("demod_railed", demod_railed);
	ar & make_nvp("squid_flux_bias", squid_flux_bias);
	ar & make_nvp("squid_current_bias", squid_current_bias);
	ar & make_nvp("squid_stage1_offset", squid_stage1_offset);
	ar & make_nvp("squid_feedback", squid_feedback);
	ar & make_nvp("routing_type", routing_type);
	ar & make_nvp("channels", channels);

	if (v >= 2) {
		ar & make_nvp("squid_state", squid_state);
		ar & make_nvp("squid_p2p", squid_p2p);
		ar & make_nvp("squid_transimpedance", squid_transimpedance);
	}
}

std::string HkMezzanineInfo::Description() const
{
	std::ostringstream s;
	s << "Mezzanine serial " << serial << " (" << part_number <<
	    ") powered " << (power ? "on" : "off") << " and " <<
	    (present ? "" : "not ") << "present";
	return s.str();
}

template <class A> void HkMezzanineInfo::serialize(A &ar, unsigned v)
{
	using namespace cereal;

	G3_CHECK_VERSION(v);

	ar & make_nvp("G3FrameObject", base_class<G3FrameObject>(this));
	ar & make_nvp("power", power);
	ar & make_nvp("present", present);
	ar & make_nvp("serial", serial);
	ar & make_nvp("part_number", part_number);
	ar & make_nvp("revision", revision);
	ar & make_nvp("currents", currents);
	ar & make_nvp("voltages", voltages);
	ar & make_nvp("modules", modules);

	if (v >= 2) {
		ar & make_nvp("temperature", temperature);
		ar & make_nvp("squid_controller_temperature", squid_controller_temperature);
		ar & make_nvp("squid_controller_power", squid_controller_power);
		ar & make_nvp("squid_heater", squid_heater);
	}
}

std::string HkBoardInfo::Description() const
{
	std::ostringstream s;
	s << "Board serial " << serial << ", set to FIR " << fir_stage <<
	    ", at " << timestamp;
	return s.str();
}

template <class A> void HkBoardInfo::serialize(A &ar, unsigned v)
{
	using namespace cereal;

	G3_CHECK_VERSION(v);

	ar & make_nvp("G3FrameObject", base_class<G3FrameObject>(this));
	ar & make_nvp("timestamp", timestamp);
	ar & make_nvp("timestamp_port", timestamp_port);
	ar & make_nvp("serial", serial);
	ar & make_nvp("fir_stage", fir_stage);

	ar & make_nvp("currents", currents);
	ar & make_nvp("voltages", voltages);
	ar & make_nvp("temperatures", temperatures);
	ar & make_nvp("mezz", mezz);

	if (v >= 2) {
		ar & make_nvp("is128x", is128x);
	}
}

G3_SERIALIZABLE_CODE(HkChannelInfo);
G3_SERIALIZABLE_CODE(HkModuleInfo);
G3_SERIALIZABLE_CODE(HkMezzanineInfo);
G3_SERIALIZABLE_CODE(HkBoardInfo);
G3_SERIALIZABLE_CODE(DfMuxHousekeepingMap);

PYBINDINGS("dfmux") {
	EXPORT_FRAMEOBJECT(HkChannelInfo, init<>(), "Mux channel status "
	  "(configuration and sensors). Usually a part of an HkModuleInfo.")
	    .def_readwrite("channel_number", &HkChannelInfo::channel_number,
               "1-indexed channel number.")
	    .def_readwrite("carrier_amplitude",
	       &HkChannelInfo::carrier_amplitude,
               "Carrier amplitude in normalized units (0-1).")
	    .def_readwrite("carrier_frequency",
	       &HkChannelInfo::carrier_frequency,
	       "Carrier frequency in standard frequency units")
	    .def_readwrite("dan_accumulator_enable",
	       &HkChannelInfo::dan_accumulator_enable,
	       "True if DAN accumulator enabled")
	    .def_readwrite("dan_feedback_enable",
	       &HkChannelInfo::dan_feedback_enable,
	       "True if DAN control of the nuller is enabled")
	    .def_readwrite("dan_streaming_enable",
	       &HkChannelInfo::dan_streaming_enable,
	       "True if streamer packets are from DAN rather than demod")
	    .def_readwrite("dan_gain", &HkChannelInfo::dan_gain,
	       "DAN gain in board units")
	    .def_readwrite("demod_frequency", &HkChannelInfo::demod_frequency,
	       "Frequency of the demodulator in standard frequency units.")
	    .def_readwrite("nuller_amplitude", &HkChannelInfo::nuller_amplitude,
	       "Nuller amplitude in normalized units (0-1)")
	    .def_readwrite("dan_railed", &HkChannelInfo::dan_railed,
	       "True if DAN has railed.")
	    .def_readwrite("state", &HkChannelInfo::state,
	       "String code (\"latched\", \"tuned\" etc.) describing the state "
	       "of this detector stored by the control software")
	    .def_readwrite("rlatched", &HkChannelInfo::rlatched,
	       "Resistance of the detector when latched in standard impedance "
	       "units as stored by the control software tuning script.")
	    .def_readwrite("rnormal", &HkChannelInfo::rnormal,
	       "Resistance of the detector when normal in standard impedance "
	       "units as stored by the control software tuning script.")
	    .def_readwrite("rfrac_achieved", &HkChannelInfo::rfrac_achieved,
	       "Achieved resistance of the detector when tuned as a fraction "
	       "of rnormal, as stored by the control software tuning script.")
	    .def_readwrite("loopgain", &HkChannelInfo::loopgain,
	       "Measured loopgain of the detector as stored by the "
	       "control software tuning script.")
	;
	register_map<std::map<int, HkChannelInfo> >("HkChannelInfoMap",
	    "Mapping of channel number (1-indexed) to channel status "
	    "information");

	EXPORT_FRAMEOBJECT(HkModuleInfo, init<>(), "Mux module status")
	    .def_readwrite("module_number", &HkModuleInfo::module_number,
	       "1-indexed module number on this mezzanine")
	    .def_readwrite("carrier_gain", &HkModuleInfo::carrier_gain,
	       "Carrier gain code, in board-specific units")
	    .def_readwrite("nuller_gain", &HkModuleInfo::nuller_gain,
	       "Nuller gain code, in board-specific units")
	    .def_readwrite("demod_gain", &HkModuleInfo::demod_gain,
	       "Demod gain code, in board-specific units")
	    .def_readwrite("carrier_railed", &HkModuleInfo::carrier_railed,
	       "True if carrier has recently hit a DAC rail")
	    .def_readwrite("nuller_railed", &HkModuleInfo::nuller_railed,
	       "True if nuller has recently hit a DAC rail")
	    .def_readwrite("demod_railed", &HkModuleInfo::demod_railed,
	       "True if demod input has recently hit one of the ADC rails")
	    .def_readwrite("squid_flux_bias", &HkModuleInfo::squid_flux_bias,
	       "Flux bias, in board-specific units (XXX)")
	    .def_readwrite("squid_current_bias", &HkModuleInfo::squid_current_bias,
	       "SQUID current bias, in board-specific units (XXX)")
	    .def_readwrite("squid_stage1_offset", &HkModuleInfo::squid_stage1_offset,
	       "Offset voltage applied to SQUID output before first-stage amp")
	    .def_readwrite("squid_p2p", &HkModuleInfo::squid_p2p,
	       "Peak-to-peak voltage difference in V-Phi curve stored by "
	       "tuning script")
	    .def_readwrite("squid_transimpedance", &HkModuleInfo::squid_transimpedance,
	       "Measured SQUID transimpedance, in resistance units")
	    .def_readwrite("squid_state", &HkModuleInfo::squid_state,
	       "Descriptive string (e.g. \"Tuned\") stored by control system "
	       "tuning script to describe SQUID state")
	    .def_readwrite("squid_feedback", &HkModuleInfo::squid_feedback,
	       "SQUID feedback mechanism employed")
	    .def_readwrite("routing_type", &HkModuleInfo::routing_type,
	       "Whether DAC are routed directly to ADCs or to the cryostat")
	    .def_readwrite("channels", &HkModuleInfo::channels,
	       "Mapping from 1-indexed channel numbers to channel housekeeping "
	       "data")
	;
	register_map<std::map<int, HkModuleInfo> >("HkModuleInfoMap",
	    "Mapping from 1-indexed module numbers to module-specific "
	    "housekeeping data");

	EXPORT_FRAMEOBJECT(HkMezzanineInfo, init<>(), "Mux mezzanine status")
	    .def_readwrite("power", &HkMezzanineInfo::power, "True if on")
	    .def_readwrite("present", &HkMezzanineInfo::present, "True if exists")
	    .def_readwrite("serial", &HkMezzanineInfo::serial,
	        "Mezzanine serial number")
	    .def_readwrite("part_number", &HkMezzanineInfo::part_number,
	        "Mezzanine part ID (usually \"MGMEZZ04\")")
	    .def_readwrite("revision", &HkMezzanineInfo::revision,
	        "Mezzanine revision number")
	    .def_readwrite("currents", &HkMezzanineInfo::currents,
	        "Dictionary of measured currents on mezzanine")
	    .def_readwrite("voltages", &HkMezzanineInfo::voltages,
	        "Dictionary of measured voltages on mezzanine")
	    .def_readwrite("modules", &HkMezzanineInfo::modules,
	        "1-indexed list of housekeeping data from readout modules on "
	        "this mezzanine")
	    .def_readwrite("temperature", &HkMezzanineInfo::temperature,
	        "Mezzanine temperature (C)")
	    .def_readwrite("squid_controller_temperature",
	        &HkMezzanineInfo::squid_controller_temperature,
	        "Measured temperature of SQUID controller board (C)")
	    .def_readwrite("squid_controller_power",
	        &HkMezzanineInfo::squid_controller_power,
	        "True if SQUID controller board powered up")
	    .def_readwrite("squid_heater", &HkMezzanineInfo::squid_heater,
	        "Power level of SQUID header control")
	;
	register_map<std::map<int, HkMezzanineInfo> >("HkMezzanineInfoMap",
	    "1-indexed mapping of mezzanine ID to mezzanine-specific "
	    "housekeeping data");

	EXPORT_FRAMEOBJECT(HkBoardInfo, init<>(), "Mux board status. Includes "
	  "both configuration and sensor readings for board generic quantities "
	  "and a list of quantities for the mezzanines.")
	    .def_readwrite("timestamp", &HkBoardInfo::timestamp,
	       "Time at which housekeeping data collected")
	    .def_readwrite("timestamp_port", &HkBoardInfo::timestamp_port,
	       "Source of timestamps on board")
	    .def_readwrite("serial", &HkBoardInfo::serial,
	       "Board serial number")
	    .def_readwrite("fir_stage", &HkBoardInfo::fir_stage,
	       "Sample rate encoded as the \"FIR Stage\". Smaller numbers are "
	       "faster and grow by factors of two with each decrement")
	    .def_readwrite("is128x", &HkBoardInfo::is128x,
		"Boolean for whether 128x firmware is running")
	    .def_readwrite("currents", &HkBoardInfo::currents,
	       "Dictionary of data from on-board current sensors")
	    .def_readwrite("voltages", &HkBoardInfo::voltages,
	       "Dictionary of data from on-board voltage sensors")
	    .def_readwrite("temperatures", &HkBoardInfo::temperatures,
	       "Dictionary of data from on-board temperature sensors (C)")
	    .def_readwrite("mezz", &HkBoardInfo::mezz,
	       "1-indexed mapping from mezzanine ID to mezzanine-specific data")
	;

	register_g3map<DfMuxHousekeepingMap>("DfMuxHousekeepingMap",
	   "Container structure for housekeeping data from all DfMux boards, "
	   "indexed by board serial number.");
}
